// @bananacool467/authtics-host/dist/cli.js
#!/usr/bin/env node
import { createServer } from "node:http";
import { createJiti } from "jiti";
import path from "node:path";
import fs from "node:fs";
import url from "node:url";
import { renderPage } from "./renderEngine.js";
let config_ = {};
let uptime = 0;
let totalRequests = 0;
let users = 0;
let usersPaused = false;
const jiti = createJiti(import.meta.url);
const routes = new Map();
const __filename = url.fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const hmrClients = new Set();
function notifyHmrClients() {
    for (const client of hmrClients) {
        try {
            client.write('data: reload\n\n');
        }
        catch {
            hmrClients.delete(client);
        }
    }
}
const MIME_TYPES = {
    '.html': 'text/html',
    '.css': 'text/css',
    '.js': 'application/javascript',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon',
    '.webp': 'image/webp',
    '.woff': 'font/woff',
    '.woff2': 'font/woff2',
    '.ttf': 'font/ttf',
    '.mp4': 'video/mp4',
    '.mp3': 'audio/mpeg',
    '.txt': 'text/plain',
    '.pdf': 'application/pdf',
};
const getCpu = () => {
    return new Promise((resolve) => {
        const start = process.cpuUsage();
        setTimeout(() => {
            const end = process.cpuUsage(start);
            const total = (end.user + end.system) / 1000;
            resolve(total);
        }, 100);
    });
};
async function getJsonBody(req) {
    return new Promise((resolve, reject) => {
        const hasBody = req.headers['content-length'] || req.headers['transfer-encoding'];
        if (!hasBody && (req.method === 'GET' || req.method === 'DELETE')) {
            return resolve({});
        }
        const chunks = [];
        const timeout = setTimeout(() => {
            req.destroy();
            reject(new Error('Request Timeout: Body took too long to arrive'));
        }, 5000);
        req.on('data', (chunk) => {
            chunks.push(chunk);
        });
        req.on('end', () => {
            clearTimeout(timeout);
            try {
                const rawData = Buffer.concat(chunks).toString();
                resolve(rawData ? JSON.parse(rawData) : {});
            }
            catch (err) {
                reject(new Error('Invalid JSON format'));
            }
        });
        req.on('error', (err) => {
            clearTimeout(timeout);
            reject(err);
        });
    });
}
export default class App {
    server;
    constructor() {
        this.server = start();
        this.post('/__authtics_host_apis__/set-users-paused', async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                usersPaused = req.body.paused;
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    message: "ok"
                }));
            }
        });
        this.post('/__authtics_host_apis__/is-users-paused', async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            res.writeHead(200, { "Content-Type": "application/json" });
            if (datMatched) {
                res.end(JSON.stringify({
                    paused: (datMatched ? false : usersPaused),
                    realInfo: usersPaused
                }));
            }
        });
        this.post('/__authtics_host_apis__/get-dat-restriction-missing', async (req, res) => {
            let datMatched = false;
            let datRestrictedEnv = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat == undefined)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                if (config_.dat !== undefined)
                    datRestrictedEnv = true;
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    enabled: datRestrictedEnv
                }));
            }
        });
        this.post('/__authtics_host_apis__/get-dat-matching', async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat == undefined)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            res.writeHead(200, { "Content-Type": "application/json" });
            res.end(JSON.stringify({
                isMatching: datMatched
            }));
        });
        this.post('/__authtics_host_apis__/get-cpu_memory', async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                const { heapUsed, heapTotal, rss } = process.memoryUsage();
                const cpuUsage = await getCpu();
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    cpuUsage,
                    heapUsed
                }));
            }
        });
        this.post('/__authtics_host_apis__/get-user-settings', async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    showError: config_.showError !== undefined ? config_.showError : true
                }));
            }
        });
        this.post("/__authtics_host_apis__/get-server-info", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    uptime, totalRequests, users
                }));
            }
        });
        this.post("/__authtics_host_apis__/restart", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({ message: "Restarting..." }));
                const { spawn } = await import('node:child_process');
                const child = spawn(process.argv[0], process.argv.slice(1), {
                    detached: true,
                    stdio: 'inherit'
                });
                child.unref();
                process.kill(process.pid, 'SIGTERM');
            }
        });
        this.post("/__authtics_host_apis__/shutdown", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({ message: "Shutting down..." }));
                this.server.close(() => {
                    process.kill(process.pid, 'SIGTERM');
                });
            }
        });
        this.post("/__authtics_host_apis__/get-is-ssr", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    isSsr: config_.ssr
                }));
            }
        });
        this.post("/__authtics_host_apis__/get-compiler", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    compiler: config_.compilerType
                }));
            }
        });
        this.post("/__authtics_host_apis__/get-host-ver", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                const packageJsonPath = path.join(__dirname, '../package.json');
                let packageJson;
                if (fs.existsSync(packageJsonPath))
                    packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf-8'));
                const packageJsonVersion = packageJson.version || "Unknown";
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    version: packageJsonVersion
                }));
            }
        });
        this.post("/__authtics_host_apis__/get-file-by-pathname", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                const pathname = req.body.pathname;
                const pagesDir = config_.autoPage === true ? path.join(process.cwd(), config_.autoPageDir || "") : undefined;
                const safePath = path.normalize(pathname).replace(/^(\.\.[\/\\])+/, '');
                let filePath = pagesDir ? path.join(pagesDir, (safePath !== "/" ? safePath : "index") + ".tsx") : undefined;
                let fileExists = filePath ? fs.existsSync(filePath) : true;
                const publicDir = config_.autoFile === true ? path.join(process.cwd(), config_.autoFileDir || "") : undefined;
                if (publicDir && !fileExists) {
                    const staticPath = path.join(publicDir, pathname);
                    const safePath = path.resolve(staticPath);
                    if (safePath.startsWith(path.resolve(publicDir))) {
                        filePath = safePath;
                        fileExists = fs.existsSync(safePath) && fs.statSync(safePath).isFile();
                    }
                }
                if (pathname.startsWith("/api/") && pagesDir != null && !fileExists) {
                    const apiFile = path.join(pagesDir, `${pathname}.ts`);
                    const apiFileTsx = path.join(pagesDir, `${pathname}.tsx`);
                    filePath = fs.existsSync(apiFile) ? apiFile : fs.existsSync(apiFileTsx) ? apiFileTsx : apiFile;
                    fileExists = filePath == apiFileTsx ? fs.existsSync(apiFileTsx) : fs.existsSync(apiFile);
                }
                if (!fileExists) {
                    filePath = "File is unresolvable. Could be because it's not found.";
                }
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    fileExists,
                    filePath: (filePath ? filePath : "Cannot find file because autoPage is disabled.")
                }));
            }
        });
        this.post("/__authtics_host_apis__/ping", async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.setHeader("Content-Type", "application/json");
                res.end(JSON.stringify({ status: "ok" }));
            }
        });
        this.post('/__authtics_host_apis__/get-connection', async (req, res) => {
            let datMatched = false;
            if (config_.dat && config_.dat == req.cookies.dat)
                datMatched = true;
            if (config_.dat && Array.isArray(config_.dat) && config_.dat.includes(req.cookies.dat))
                datMatched = true;
            if (datMatched) {
                res.writeHead(200, { "Content-Type": "application/json" });
                res.end(JSON.stringify({
                    httpVersion: req.httpVersion
                }));
            }
        });
    }
    get(routePath, handler) {
        routes.set(`GET:${routePath}`, handler);
    }
    registerApp(route) {
        routes.set(`APP_TSX`, route);
    }
    post(routePath, handler) {
        routes.set(`POST:${routePath}`, handler);
    }
    put(routePath, handler) {
        routes.set(`PUT:${routePath}`, handler);
    }
    delete(routePath, handler) {
        routes.set(`DELETE:${routePath}`, handler);
    }
    remove(method, routePath) {
        if (routes.has(`${method}:${routePath}`))
            routes.delete(`${method}:${routePath}`);
    }
}
async function start() {
    let config = {};
    try {
        const configPath = path.join(process.cwd(), "authtics_host.config.ts");
        config = await jiti.import(configPath, { default: true });
        config_ = config;
    }
    catch {
    }
    const PORT = config.port ?? 3000;
    const showPanel = config.devMode !== false;
    const autoPaging = config.autoPage !== false;
    const compiler = config.compilerType === 'babel' ? 'babel' : 'esbuild';
    const hotReload = config.hotReload === true;
    const pagesDir = config.autoPage === true ? path.join(process.cwd(), config.autoPageDir || "") : undefined;
    const publicDir = config.autoFile === true ? path.join(process.cwd(), config.autoFileDir || "") : undefined;
    const apiDir = pagesDir ? path.join(pagesDir, "api") : undefined;
    const server = createServer(async (req, res) => {
        const { url = "/", method } = req;
        const actuallyShowPanel = showPanel;
        req.ip = (req.headers["x-real-ip"] ||
            req.headers["x-forwarded-for"]?.split(",")[0].trim() ||
            req.socket.remoteAddress ||
            undefined);
        req.ipType = req.ip?.includes(":") ? 6 : req.ip?.includes(".") ? 4 : 0;
        req.ua = req.headers["user-agent"] || undefined;
        let ua = req.headers["user-agent"] || "";
        let browserType = "Unknown";
        let browserVersion = "Unknown";
        if (ua.includes("Chrome") && !ua.includes("Edg")) {
            browserType = "Chrome";
            browserVersion = ua.match(/Chrome\/([\d.]+)/)?.[1] || "Unknown";
        }
        else if (ua.includes("Firefox")) {
            browserType = "Firefox";
            browserVersion = ua.match(/Firefox\/([\d.]+)/)?.[1] || "Unknown";
        }
        else if (ua.includes("Safari") && !ua.includes("Chrome")) {
            browserType = "Safari";
            browserVersion = ua.match(/Version\/([\d.]+)/)?.[1] || "Unknown";
        }
        else if (ua.includes("Edg")) {
            browserType = "Edge";
            browserVersion = ua.match(/Edg\/([\d.]+)/)?.[1] || "Unknown";
        }
        let osType = "Unknown";
        let osVersion = "Unknown";
        if (ua.includes("Windows NT")) {
            osType = "Windows";
            osVersion = ua.match(/Windows NT ([\d.]+)/)?.[1] || "Unknown";
        }
        else if (ua.includes("Mac OS X")) {
            osType = "macOS";
            osVersion = ua.match(/Mac OS X ([\d_]+)/)?.[1]?.replace(/_/g, ".") || "Unknown";
        }
        else if (ua.includes("Android")) {
            osType = "Android";
            osVersion = ua.match(/Android ([\d.]+)/)?.[1] || "Unknown";
        }
        else if (ua.includes("iPhone") || ua.includes("iPad")) {
            osType = "iOS";
            osVersion = ua.match(/OS ([\d_]+)/)?.[1]?.replace(/_/g, ".") || "Unknown";
        }
        else if (ua.includes("Linux")) {
            osType = "Linux";
        }
        req.os = {
            type: osType,
            version: osVersion
        };
        req.browser = {
            type: browserType,
            version: browserVersion
        };
        req.sentTime = Date.now();
        req.body = {};
        try {
            req.body = await getJsonBody(req);
        }
        catch (e) {
            req.body = {};
        }
        req.cookies = {};
        const raw = req.headers.cookie;
        if (raw) {
            raw.split(";").forEach(pair => {
                const [key, val] = pair.trim().split("=");
                req.cookies[key] = val;
            });
        }
        req.dat = req.cookies.dat || undefined;
        req.bodySize = JSON.stringify(req.body).length;
        req.bodyType = req.headers['content-type'] || 'text/plain';
        res.redirect = (url) => {
            res.writeHead(302, {
                'Location': url
            });
            res.end();
        };
        res.html = (content) => {
            res.setHeader("Content-Type", "text/html");
            res.end(content);
        };
        res.json = (content) => {
            res.setHeader("Content-Type", "application/json");
            res.end(JSON.stringify(content));
        };
        res.text = (content) => {
            res.setHeader("Content-Type", "text/plain");
            res.end(content);
        };
        res.file = (path_) => {
            const path__ = path.join(process.cwd(), path_);
            if (fs.existsSync(path__)) {
                const ext = path.extname(path__);
                const mime = MIME_TYPES[ext] ?? 'application/octet-stream';
                res.setHeader("Content-Type", mime);
                res.end(fs.readFileSync(path__));
            }
        };
        res.status = (code) => {
            res.statusCode = code;
            return res;
        };
        res.reload = function () {
            res.writeHead(302, { Location: req.url + '/' });
            res.end();
        };
        res.setCookie = function (name, value) {
            res.setHeader("Set-Cookie", `${name}=${value}; Path=/`);
        };
        res.rmvCookie = function (name) {
            res.setHeader("Set-Cookie", `${name}=; Path=/; Max-Age=0`);
        };
        let pathname = url.split("?")[0].split("#")[0];
        const safePath = path.normalize(pathname).replace(/^(\.\.[\/\\])+/, '');
        pathname = safePath;
        if (pathname !== "/" && pathname.endsWith("/"))
            pathname = pathname.slice(0, -1);
        if (!pathname.includes('/__authtics') && !pathname.includes('/hmr')) {
            totalRequests++;
        }
        const routeKey = `${method}:${pathname}`;
        if (routes.has(routeKey)) {
            if (Array.isArray(config_.dat) && !config_.dat.includes(req.cookies.dat) && usersPaused && !routeKey.endsWith('.js')) {
                if (autoPaging) {
                    const pausedTsxFile = pagesDir ? path.join(pagesDir, "_paused.tsx") : undefined;
                    if (fs.existsSync(pausedTsxFile || "")) {
                        try {
                            let html = await renderPage(pausedTsxFile || "", url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                                routes.set(`GET:${p}`, (req, res) => {
                                    res.setHeader("Content-Type", "application/javascript");
                                    res.end(f);
                                });
                            }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                            html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                            res.setHeader("Content-Type", "text/html");
                            return res.end(html);
                        }
                        catch (e) {
                            console.error("[authtics] renderPage error:", e.message);
                            res.writeHead(500);
                            return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                        }
                    }
                }
                if (config.hasPausedRoute) {
                    if (!autoPaging) {
                        const appTsxPath = routes.get("APP_TSX");
                        const appPath = path.join(process.cwd(), appTsxPath);
                        if (fs.existsSync(appPath)) {
                            try {
                                let html = await renderPage(appPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                                    routes.set(`GET:${p}`, (req, res) => {
                                        res.setHeader("Content-Type", "application/javascript");
                                        res.end(f);
                                    });
                                }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                                html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                                res.setHeader("Content-Type", "text/html");
                                return res.end(html);
                            }
                            catch (e) {
                                console.error("[authtics] renderPage error:", e.message);
                                res.writeHead(500);
                                return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                            }
                        }
                    }
                }
                res.writeHead(200, {
                    "Content-Type": "application/json"
                });
                return res.end(JSON.stringify({
                    message: "unhandled request"
                }));
            }
            if (config_.dat !== req.cookies.dat && usersPaused && !routeKey.endsWith('.js')) {
                if (autoPaging) {
                    const pausedTsxFile = pagesDir ? path.join(pagesDir, "_paused.tsx") : undefined;
                    if (fs.existsSync(pausedTsxFile || "")) {
                        try {
                            let html = await renderPage(pausedTsxFile || "", url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                                routes.set(`GET:${p}`, (req, res) => {
                                    res.setHeader("Content-Type", "application/javascript");
                                    res.end(f);
                                });
                            }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                            html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                            res.setHeader("Content-Type", "text/html");
                            return res.end(html);
                        }
                        catch (e) {
                            console.error("[authtics] renderPage error:", e.message);
                            res.writeHead(500);
                            return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                        }
                    }
                }
                if (config.hasPausedRoute) {
                    if (!autoPaging) {
                        const appTsxPath = routes.get("APP_TSX");
                        const appPath = path.join(process.cwd(), appTsxPath);
                        if (fs.existsSync(appPath)) {
                            try {
                                let html = await renderPage(appPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                                    routes.set(`GET:${p}`, (req, res) => {
                                        res.setHeader("Content-Type", "application/javascript");
                                        res.end(f);
                                    });
                                }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                                html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                                res.setHeader("Content-Type", "text/html");
                                return res.end(html);
                            }
                            catch (e) {
                                console.error("[authtics] renderPage error:", e.message);
                                res.writeHead(500);
                                return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                            }
                        }
                    }
                }
                res.writeHead(200, {
                    "Content-Type": "application/json"
                });
                return res.end(JSON.stringify({
                    message: "unhandled request"
                }));
            }
            const handler = routes.get(routeKey);
            if (typeof handler === 'string') {
                return;
            }
            else if (typeof handler === 'function') {
                return handler(req, res);
            }
        }
        if (pathname === '/hmr') {
            res.writeHead(200, {
                'Content-Type': 'text/event-stream',
                'Cache-Control': 'no-cache',
                'Connection': 'keep-alive',
                'Access-Control-Allow-Origin': '*',
            });
            res.write('data: connected\n\n');
            hmrClients.add(res);
            users++;
            req.on('close', () => { hmrClients.delete(res); users--; });
            return;
        }
        if (Array.isArray(config_.dat) && !config_.dat.includes(req.cookies.dat) && usersPaused) {
            if (autoPaging) {
                const pausedTsxFile = pagesDir ? path.join(pagesDir, "_paused.tsx") : undefined;
                if (fs.existsSync(pausedTsxFile || "")) {
                    try {
                        let html = await renderPage(pausedTsxFile || "", url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                            routes.set(`GET:${p}`, (req, res) => {
                                res.setHeader("Content-Type", "application/javascript");
                                res.end(f);
                            });
                        }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                        html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                        res.setHeader("Content-Type", "text/html");
                        return res.end(html);
                    }
                    catch (e) {
                        console.error("[authtics] renderPage error:", e.message);
                        res.writeHead(500);
                        return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                    }
                }
            }
            if (config.hasPausedRoute) {
                if (!autoPaging) {
                    const appTsxPath = routes.get("APP_TSX");
                    const appPath = path.join(process.cwd(), appTsxPath);
                    if (fs.existsSync(appPath)) {
                        try {
                            let html = await renderPage(appPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                                routes.set(`GET:${p}`, (req, res) => {
                                    res.setHeader("Content-Type", "application/javascript");
                                    res.end(f);
                                });
                            }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                            html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                            res.setHeader("Content-Type", "text/html");
                            return res.end(html);
                        }
                        catch (e) {
                            console.error("[authtics] renderPage error:", e.message);
                            res.writeHead(500);
                            return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                        }
                    }
                }
            }
            res.writeHead(200, {
                "Content-Type": "application/json"
            });
            return res.end(JSON.stringify({
                message: "unhandled request"
            }));
        }
        if (config.dat !== req.cookies.dat && usersPaused) {
            if (autoPaging) {
                const pausedTsxFile = pagesDir ? path.join(pagesDir, "_paused.tsx") : undefined;
                if (fs.existsSync(pausedTsxFile || "")) {
                    try {
                        let html = await renderPage(pausedTsxFile || "", url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                            routes.set(`GET:${p}`, (req, res) => {
                                res.setHeader("Content-Type", "application/javascript");
                                res.end(f);
                            });
                        }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                        html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                        res.setHeader("Content-Type", "text/html");
                        return res.end(html);
                    }
                    catch (e) {
                        console.error("[authtics] renderPage error:", e.message);
                        res.writeHead(500);
                        return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                    }
                }
            }
            if (config.hasPausedRoute) {
                if (!autoPaging) {
                    const appTsxPath = routes.get("APP_TSX");
                    const appPath = path.join(process.cwd(), appTsxPath);
                    if (fs.existsSync(appPath)) {
                        try {
                            let html = await renderPage(appPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                                routes.set(`GET:${p}`, (req, res) => {
                                    res.setHeader("Content-Type", "application/javascript");
                                    res.end(f);
                                });
                            }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                            html = html.replace("<script>", "<script>\nwindow.__paused__ = true;");
                            res.setHeader("Content-Type", "text/html");
                            return res.end(html);
                        }
                        catch (e) {
                            console.error("[authtics] renderPage error:", e.message);
                            res.writeHead(500);
                            return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                        }
                    }
                }
            }
            res.writeHead(200, {
                "Content-Type": "application/json"
            });
            return res.end(JSON.stringify({
                message: "unhandled request"
            }));
        }
        if (publicDir && fs.existsSync(publicDir)) {
            const staticPath = path.join(publicDir, pathname);
            const safePath = path.resolve(staticPath);
            if (safePath.startsWith(path.resolve(publicDir)) && fs.existsSync(safePath) && fs.statSync(safePath).isFile()) {
                const ext = path.extname(safePath).toLowerCase();
                const mime = MIME_TYPES[ext] ?? 'application/octet-stream';
                res.setHeader("Content-Type", mime);
                return fs.createReadStream(safePath).pipe(res);
            }
        }
        if (pathname.startsWith("/api/") && pagesDir != null) {
            const apiFile = path.join(pagesDir, `${pathname}.ts`);
            const apiFileTsx = path.join(pagesDir, `${pathname}.tsx`);
            const resolvedApi = fs.existsSync(apiFile) ? apiFile : fs.existsSync(apiFileTsx) ? apiFileTsx : null;
            if (resolvedApi && fs.existsSync(resolvedApi)) {
                try {
                    const apiModule = await jiti.import(resolvedApi, { default: true });
                    const handler = apiModule.default ?? apiModule.handler ?? null;
                    if (typeof handler !== 'function') {
                        res.writeHead(500);
                        return res.end(`[authtics] API route ${pathname} must export a default function`);
                    }
                    return handler(req, res);
                }
                catch (e) {
                    console.error(`[authtics] API route error (${pathname}):`, e.message);
                    res.writeHead(500);
                    return res.end(JSON.stringify({ error: e.message }));
                }
            }
            const notFoundPath = path.join(__dirname, '404.js');
            if (fs.existsSync(notFoundPath)) {
                try {
                    let html = await renderPage(notFoundPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                        routes.set(`GET:${p}`, (req, res) => {
                            res.setHeader("Content-Type", "application/javascript");
                            res.end(f);
                        });
                    }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                    res.setHeader("Content-Type", "text/html");
                    res.writeHead(404);
                    html = html.replace("<script>", "<script>\nwindow.__status_code__ = 404;");
                    res.end(html);
                }
                catch (e) {
                    res.writeHead(404);
                    res.end(`<html><head></head><body><center><b>404</b> | Not found: ${url}</center></body></html>`);
                }
            }
            else {
                res.writeHead(404);
                res.end(`<html><head></head><body><center><b>404</b> | Not found: ${url}</center></body></html>`);
            }
        }
        const urlPath = pathname === "/" ? "/index" : pathname;
        const tsxPath = pagesDir ? path.join(pagesDir, `${urlPath}.tsx`) : undefined;
        const htmlPath = pagesDir ? path.join(pagesDir, `${urlPath}.html`) : undefined;
        if (tsxPath && fs.existsSync(tsxPath)) {
            try {
                const html = await renderPage(tsxPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                    routes.set(`GET:${p}`, (req, res) => {
                        res.setHeader("Content-Type", "application/javascript");
                        res.end(f);
                    });
                }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                res.setHeader("Content-Type", "text/html");
                return res.end(html);
            }
            catch (e) {
                console.error("[authtics] renderPage error:", e.message);
                res.writeHead(500);
                return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
            }
        }
        if (htmlPath && fs.existsSync(htmlPath)) {
            res.setHeader("Content-Type", "text/html");
            return res.end(fs.readFileSync(htmlPath, "utf-8"));
        }
        if (routes.has("APP_TSX")) {
            const appTsxPath = routes.get("APP_TSX");
            const appPath = path.join(process.cwd(), appTsxPath);
            if (fs.existsSync(appPath)) {
                try {
                    const html = await renderPage(appPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                        routes.set(`GET:${p}`, (req, res) => {
                            res.setHeader("Content-Type", "application/javascript");
                            res.end(f);
                        });
                    }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                    res.setHeader("Content-Type", "text/html");
                    return res.end(html);
                }
                catch (e) {
                    console.error("[authtics] renderPage error:", e.message);
                    res.writeHead(500);
                    return res.end(`<html><body><pre>500 | Render error:\n${e.message}</pre></body></html>`);
                }
            }
        }
        const notFoundPath = path.join(__dirname, '404.js');
        if (fs.existsSync(notFoundPath)) {
            try {
                let html = await renderPage(notFoundPath, url, actuallyShowPanel, compiler, hotReload, autoPaging, config.autoPageDir, (routes.get("APP_TSX") || ""), (f, p) => {
                    routes.set(`GET:${p}`, (req, res) => {
                        res.setHeader("Content-Type", "application/javascript");
                        res.end(f);
                    });
                }, (p) => { routes.delete(`GET:${p}`); }, req.cookies.dat, config_.dat);
                res.setHeader("Content-Type", "text/html");
                res.writeHead(404);
                html = html.replace("<script>", "<script>\nwindow.__status_code__ = 404;");
                res.end(html);
            }
            catch (e) {
                res.writeHead(404);
                res.end(`<html><head></head><body><center><b>404</b> | Not found: ${url}</center></body></html>`);
            }
        }
        else {
            res.writeHead(404);
            res.end(`<html><head></head><body><center><b>404</b> | Not found: ${url}</center></body></html>`);
        }
    });
    server.listen(PORT, () => {
        setInterval(() => {
            uptime++;
        }, 1000);
    });
    if (config.hotReload === true) {
        const watchDirs = [pagesDir || path.join(routes.get("APP_TSX"), "../")];
        const devPanelDir = path.join(__dirname, 'devPanel.js');
        const _404Dir = path.join(__dirname, '404.js');
        if (fs.existsSync(devPanelDir))
            watchDirs.push(devPanelDir);
        for (const dir of watchDirs) {
            if (dir && !fs.existsSync(dir))
                continue;
            if (dir) {
                fs.watch(dir, { recursive: true }, (eventType, filename) => {
                    if (!filename)
                        return;
                    notifyHmrClients();
                });
            }
        }
        if (fs.existsSync(_404Dir))
            watchDirs.push(_404Dir);
        for (const dir of watchDirs) {
            if (dir && !fs.existsSync(dir))
                continue;
            if (dir) {
                fs.watch(dir, { recursive: true }, (eventType, filename) => {
                    if (!filename)
                        return;
                    notifyHmrClients();
                });
            }
        }
    }
    return server;
}
