// @bananacool467/authtics-host/dist/devPanel.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useState, useEffect, useRef } from 'react';
import { ShieldCheck, CircleX, ChevronDown, ChevronUp, Bug, AlertTriangle, Info, ArrowLeft, Wifi, Terminal, Route, Globe, ChevronRight, ChevronLeft, Server, TriangleAlert, Eye, Zap } from 'lucide-react';
function getCallerSource() {
    try {
        const stack = new Error().stack ?? '';
        const lines = stack.split('\n');
        for (let i = 3; i < lines.length; i++) {
            const line = lines[i];
            if (line.trim() === '')
                continue;
            const match = line.match(/\(?(https?:\/\/[^)]+)\)?$/) ??
                line.match(/at\s+(https?:\/\/\S+)/);
            if (match) {
                const url = match[1];
                const parts = url.replace(/\?.*/, '').split('/');
                const file = parts[parts.length - 1];
                const colonLine = url.match(/:(\d+):\d+$/);
                return colonLine ? `${file.split(':')[0]}:${colonLine[1]}` : file;
            }
        }
    }
    catch { }
    return '';
}
const LOG_TYPES = ['Log', 'Info', 'Warn', 'Error', 'Debug'];
let fetchId = 0;
const TYPE_STYLE = {
    Error: { bg: '#fff0f0', color: '#e53e3e', icon: _jsx(CircleX, { size: 11 }) },
    Warn: { bg: '#fffbea', color: '#d97706', icon: _jsx(AlertTriangle, { size: 11 }) },
    Log: { bg: '#ffffff', color: '#888', textColor: '#333', icon: null },
    Info: { bg: '#eff6ff', color: '#3b82f6', icon: _jsx(Info, { size: 11 }) },
    Debug: { bg: '#f0fdf4', color: '#16a34a', icon: _jsx(Bug, { size: 11 }) },
};
const FILTER_COLORS = {
    Error: '#e53e3e',
    Warn: '#d97706',
    Log: '#888',
    Info: '#3b82f6',
    Debug: '#16a34a',
};
const TruncatedText = ({ content, id }) => {
    const maxLength = 100;
    const truncated = content.length > maxLength
        ? content.slice(0, maxLength) + "..."
        : content;
    return _jsx("div", { id: id, children: truncated });
};
function Sparkline({ data, color, label, unit }) {
    const W = 280, H = 48;
    const max = Math.max(...data, 1);
    const points = data.map((v, i) => {
        const x = (i / (data.length - 1)) * W;
        const y = H - (v / max) * (H - 4) - 2;
        return `${x},${y}`;
    }).join(' ');
    const latest = data[data.length - 1] ?? 0;
    return (_jsxs("div", { style: { background: '#fff', borderRadius: '8px', border: '1px solid #eee', padding: '8px 10px' }, children: [_jsxs("div", { style: { display: 'flex', justifyContent: 'space-between', alignItems: 'center', marginBottom: '4px' }, children: [_jsx("span", { style: { fontSize: '10px', color: '#888' }, children: label }), _jsxs("span", { style: { fontSize: '11px', fontFamily: 'monospace', color }, children: [latest, unit] })] }), _jsxs("svg", { width: W, height: H, style: { display: 'block', overflow: 'visible' }, children: [_jsx("defs", { children: _jsxs("linearGradient", { id: `grad-${label}`, x1: "0", y1: "0", x2: "0", y2: "1", children: [_jsx("stop", { offset: "0%", stopColor: color, stopOpacity: "0.18" }), _jsx("stop", { offset: "100%", stopColor: color, stopOpacity: "0.01" })] }) }), data.length > 1 && (_jsxs(_Fragment, { children: [_jsx("polygon", { points: `0,${H} ${points} ${W},${H}`, fill: `url(#grad-${label})` }), _jsx("polyline", { points: points, fill: "none", stroke: color, strokeWidth: "1.5", strokeLinejoin: "round", strokeLinecap: "round" }), (() => {
                                const last = points.split(' ').pop().split(',');
                                return _jsx("circle", { cx: last[0], cy: last[1], r: "2.5", fill: color });
                            })()] }))] })] }));
}
function installNetworkInterceptor(onFetch, setFetch, onEventSource, setEventSource, onWebSocket, setWebSocket, onResource, setResource, rmvResource, onBytesClient, onBytesServer) {
    if (window.__authtics_network_intercepted__)
        return;
    window.__authtics_network_intercepted__ = true;
    const originalFetch = window.fetch.bind(window);
    window.fetch = async (...args) => {
        fetchId++;
        const savedFetchId = fetchId;
        const url = typeof args[0] === 'string' ? args[0] : args[0].url;
        const method = (args[1]?.method ?? 'GET').toUpperCase();
        if (url.startsWith('/__authtics_host_apis__/') || url === "/__authtics_hmr__") {
            return originalFetch(...args);
        }
        const body = args[1]?.body;
        const bodyBytes = body ? new Blob([body]).size : 0;
        const headerStr = JSON.stringify(args[1]?.headers ?? {});
        const sentBytes = bodyBytes + url.length + headerStr.length;
        onBytesClient(sentBytes);
        onFetch({ url, method, status: null, id: savedFetchId, response: null, sent: body ? JSON.stringify(body, null, 2).replace('"{', '{')
                .replace('}"', '}')
                .replaceAll('\\"', '"') : null });
        try {
            const resp = await originalFetch(...args);
            const clone = resp.clone();
            setFetch({ url, method, status: resp.status, id: savedFetchId, response: await resp.text(), sent: body ? JSON.stringify(body, null, 2).replace('"{', '{')
                    .replace('}"', '}')
                    .replaceAll('\\"', '"') : null });
            clone.arrayBuffer().then(buf => {
                onBytesServer(buf.byteLength);
            }).catch(() => { });
            return resp;
        }
        catch (e) {
            onFetch({ url, method, status: 0, id: savedFetchId, response: null, sent: body ? JSON.stringify(body, null, 2).replace(/[{}/"]/g, "") : null });
            throw e;
        }
    };
    const OriginalWebSocket = window.WebSocket;
    window.WebSocket = function (url, protocols) {
        fetchId++;
        const savedId = fetchId;
        const ws = new OriginalWebSocket(url, protocols);
        if (url.includes('__authtics'))
            return ws;
        onWebSocket({
            url,
            status: 'Connecting...',
            id: savedId,
            type: "WebSocket"
        });
        ws.addEventListener('open', () => {
            setWebSocket({
                url,
                status: 'Connected',
                id: savedId,
                type: "WebSocket"
            });
        });
        ws.addEventListener('close', () => {
            setWebSocket({
                url,
                status: 'Disconnected',
                id: savedId,
                type: "WebSocket"
            });
        });
        ws.addEventListener('message', (event) => {
            const bytes = new Blob([event.data]).size;
            onBytesServer(bytes);
            setWebSocket({
                url,
                status: 'Connected',
                id: savedId,
                type: "WebSocket"
            });
        });
        ws.addEventListener('error', () => {
            const status = ws.readyState === OriginalWebSocket.CLOSED
                ? 'Disconnected'
                : 'Connection Failed';
            setWebSocket({
                url,
                status,
                id: savedId,
                type: "WebSocket"
            });
        });
        const OriginalSend = ws.send;
        ws.send = function (data) {
            let bytes = 0;
            if (typeof data === 'string') {
                bytes = new Blob([data]).size;
            }
            else if (data instanceof Blob) {
                bytes = data.size;
            }
            else if (data instanceof ArrayBuffer || data instanceof SharedArrayBuffer) {
                bytes = data.byteLength;
            }
            else if (ArrayBuffer.isView(data)) {
                bytes = data.byteLength;
            }
            onBytesClient(bytes);
            OriginalSend.call(this, data);
        };
        return ws;
    };
    Object.defineProperty(window.WebSocket, 'prototype', {
        value: OriginalWebSocket.prototype,
        writable: false
    });
    Object.assign(window.WebSocket, OriginalWebSocket);
    const OriginalEventSource = window.EventSource;
    window.EventSource = function (url, eventSourceInitDict) {
        fetchId++;
        const savedId = fetchId;
        const es = new OriginalEventSource(url, eventSourceInitDict);
        if (url.includes('__authtics'))
            return es;
        onEventSource({
            url,
            status: 'Connecting...',
            id: savedId,
            type: "EventSource"
        });
        es.addEventListener('open', () => {
            setEventSource({
                url,
                status: 'Connected',
                id: savedId,
                type: "EventSource"
            });
        });
        es.addEventListener('message', (event) => {
            const bytes = new Blob([event.data]).size;
            onBytesServer(bytes);
            setEventSource({
                url,
                status: 'Connected',
                id: savedId,
                type: "EventSource"
            });
        });
        es.addEventListener('error', () => {
            const status = es.readyState === OriginalEventSource.CLOSED
                ? 'Disconnected'
                : 'Connection Failed';
            setEventSource({
                url,
                status: status,
                id: savedId,
                type: "EventSource"
            });
        });
        return es;
    };
    window.EventSource.prototype = OriginalEventSource.prototype;
    Object.assign(window.EventSource, OriginalEventSource);
    let usedResourcesSlots = [];
    let beforeResources = [];
    let currentResources = [];
    setInterval(() => {
        beforeResources = currentResources;
        currentResources = [];
        const queryImg = document.querySelectorAll('img');
        const queryScript = document.querySelectorAll('script');
        const queryLink = document.querySelectorAll('link');
        queryImg.forEach((img) => {
            if (img.src) {
                fetchId++;
                let savedFetchId = fetchId;
                if (!usedResourcesSlots.includes({ url: img.src, type: 'img', id: savedFetchId }))
                    onResource({ url: img.src, type: 'img', id: savedFetchId });
                currentResources.push({ url: img.src, type: 'img', id: savedFetchId });
                if (usedResourcesSlots.includes({ url: img.src, type: 'img', id: savedFetchId }))
                    setResource({ url: img.src, type: 'img', id: savedFetchId });
                if (!usedResourcesSlots.includes({ url: img.src, type: 'img', id: savedFetchId }))
                    usedResourcesSlots.push({ url: img.src, type: 'img', id: savedFetchId });
            }
        });
        queryScript.forEach((script) => {
            if (script.src) {
                fetchId++;
                let savedFetchId = fetchId;
                if (!usedResourcesSlots.includes({ url: script.src, type: 'script', id: savedFetchId }))
                    onResource({ url: script.src, type: 'script', id: savedFetchId });
                currentResources.push({ url: script.src, type: 'script', id: savedFetchId });
                if (usedResourcesSlots.includes({ url: script.src, type: 'script', id: savedFetchId }))
                    setResource({ url: script.src, type: 'script', id: savedFetchId });
                if (!usedResourcesSlots.includes({ url: script.src, type: 'script', id: savedFetchId }))
                    usedResourcesSlots.push({ url: script.src, type: 'script', id: savedFetchId });
            }
        });
        queryLink.forEach((link) => {
            if (link.href) {
                fetchId++;
                let savedFetchId = fetchId;
                if (!usedResourcesSlots.includes({ url: link.href, type: 'link', id: savedFetchId }))
                    onResource({ url: link.href, type: 'link', id: savedFetchId });
                currentResources.push({ url: link.href, type: 'link', id: savedFetchId });
                if (usedResourcesSlots.includes({ url: link.href, type: 'link', id: savedFetchId }))
                    setResource({ url: link.href, type: 'link', id: savedFetchId });
                if (!usedResourcesSlots.includes({ url: link.href, type: 'link', id: savedFetchId }))
                    usedResourcesSlots.push({ url: link.href, type: 'link', id: savedFetchId });
            }
        });
        beforeResources.forEach((resource) => {
            if (!currentResources.includes(resource)) {
                rmvResource(resource);
            }
        });
    }, 250);
}
function installConsoleInterceptor() {
    if (window.__authtics_intercepted__)
        return;
    window.__authtics_intercepted__ = true;
    LOG_TYPES.forEach(type => {
        const t = type.toLowerCase();
        const original = console[t].bind(console);
        console[t] = (...args) => {
            original(...args);
            window.dispatchEvent(new CustomEvent('authtics:log', {
                detail: {
                    type,
                    message: args.map(a => typeof a === 'object' ? JSON.stringify(a, null, 2) : String(a)).join(' '),
                    time: Date.now(),
                    source: getCallerSource(),
                }
            }));
        };
    });
}
function SecuredTab() {
    return (_jsxs("div", { style: { display: 'flex', justifyContent: 'center', alignItems: 'center', flexDirection: 'column', padding: '8px 10px', textAlign: 'center', gap: '8px' }, children: [_jsx(CircleX, { size: 32, color: 'red' }), _jsx("span", { style: { color: 'red', fontSize: 24 }, children: "This Tab Is Currently Not Accessible" }), _jsx("span", { style: { color: 'red', fontSize: 12 }, children: "To access this tab, you need to set \"dat\" in the Authtics Host config to set restriction and then restart." })] }));
}
function PanelButton({ onClick, onClick2, children, style, fullWidth, disabled }) {
    const [hovered, setHovered] = useState(false);
    const [pressed, setPressed] = useState(false);
    return (_jsx("div", { onClick: (e) => {
            if (onClick && !disabled) {
                onClick();
            }
            else if (onClick2 && !disabled) {
                onClick2(e);
            }
        }, onMouseEnter: () => setHovered(true), onMouseLeave: () => { setHovered(false); setPressed(false); }, onMouseDown: () => setPressed(true), onMouseUp: () => setPressed(false), style: {
            cursor: 'pointer',
            borderRadius: '10px',
            padding: '6px 10px',
            fontSize: '11px',
            display: 'flex',
            alignItems: 'center',
            gap: '5px',
            width: fullWidth ? '100%' : undefined,
            boxSizing: 'border-box',
            userSelect: 'none',
            background: pressed ? 'rgba(0,0,0,0.12)' : hovered ? 'rgba(0,0,0,0.06)' : 'rgba(0,0,0,0.03)',
            transition: 'background 0.15s ease',
            ...style,
        }, children: children }));
}
const MAX_POINTS = 20;
const generateToken = () => {
    return crypto.randomUUID().replace(/-/g, '').substring(0, 16);
};
export default function DevPanel() {
    const [logs, setLogs] = useState([]);
    const [open, setOpen] = useState(false);
    const [errorBoxOpen, setErrorBoxOpen] = useState(false);
    const [errorMsgs, setErrorMsgs] = useState([]);
    const [selectedError, setSelectedError] = useState(1);
    const [consoleOpen, setConsoleOpen] = useState(false);
    const [networkOpen, setNetworkOpen] = useState(false);
    const [routeOpen, setRouteOpen] = useState(false);
    const [webpageInfoOpen, setWebpageInfoOpen] = useState(false);
    const [clientInfoOpen, setClientInfoOpen] = useState(false);
    const [aboutHostOpen, setAboutHostOpen] = useState(false);
    const [serverInfoOpen, setServerInfoOpen] = useState(false);
    const [hasIssue, setHasIssue] = useState(false);
    const [ping__, setPing__] = useState(0);
    const [networkFetches, setNetworkFetches] = useState([]);
    const [eventSources, setEventSources] = useState([]);
    const [usersPausedRealInfo, setUsersPausedRealInfo] = useState(false);
    const [webSockets, setWebSockets] = useState([]);
    const [resourcesFound, setResourcesFound] = useState([]);
    const [status, setStatus] = useState(200);
    const [pathName, setPathName] = useState(undefined);
    const [hostName, setHostName] = useState(undefined);
    const [protocol, setProtocol] = useState(undefined);
    const [path, setPath] = useState(undefined);
    const [isSsr, setIsSsr] = useState(false);
    const datRestrictionMissingRef = useRef(true);
    const cpuUsage = useRef(0);
    const heapUsage = useRef(0);
    const [datMatch, setDatMatch] = useState(false);
    const [compilerType, setCompilerType] = useState("?");
    const [httpVersion, setHttpVersion] = useState("v?.?");
    const [errorBoxOpenable, setErrorBoxOpenable] = useState(false);
    const [uptime, setUptime] = useState("00:00:00:00");
    const [requests, setRequests] = useState(0);
    const [hostVersion, setHostVersion] = useState("Loading...");
    const [users, setUsers] = useState(0);
    const [token, setToken] = useState(null);
    const [tokenLength, setTokenLength] = useState(0);
    const [tokenHidden, setTokenHidden] = useState(true);
    const fpsSpanRef = useRef(null);
    const frameCount = useRef(0);
    const frames = useRef(0);
    const consoleScrollRef = useRef(null);
    const [isAtBottom, setIsAtBottom] = useState(true);
    const [datRestrictionMissing, setDatRestrictionMissing] = useState(true);
    const cookie = {
        set: (k, v) => document.cookie = `${k}=${v}; path=/`,
        get: (k) => Object.fromEntries(document.cookie.split(";")
            .filter(Boolean)
            .map(c => c.trim().split("=")))[k],
        del: (k) => document.cookie = `${k}=; path=/; max-age=0`
    };
    const [doc, setDoc] = useState(null);
    const [filters, setFilters] = useState({
        Log: true, Info: true, Warn: true, Error: true, Debug: false
    });
    const errorCount = logs.filter(l => l.type === 'Error').length;
    const warnCount = logs.filter(l => l.type === 'Warn').length;
    const filteredLogs = logs.filter(l => filters[l.type]);
    const handleScroll = () => {
        if (consoleScrollRef.current) {
            const { scrollTop, scrollHeight, clientHeight } = consoleScrollRef.current;
            const reachedBottom = scrollHeight - scrollTop - clientHeight < 5;
            setIsAtBottom(reachedBottom);
        }
    };
    useEffect(() => {
        if (isAtBottom && consoleScrollRef.current) {
            consoleScrollRef.current.scrollTop = consoleScrollRef.current.scrollHeight;
        }
    }, [filteredLogs, isAtBottom]);
    const loop = () => {
        frameCount.current++;
        requestAnimationFrame(loop);
    };
    useEffect(() => {
        const getPath = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-file-by-pathname', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ pathname: window.location.pathname, })
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setPath(data.filePath);
            }
        };
        const getDatMatching = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-dat-matching', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setDatMatch(data.isMatching);
                const resp3 = await fetch('/__authtics_host_apis__/is-users-paused', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({})
                });
                let paused = false;
                if (resp3.status === 200) {
                    const data = await resp3.json();
                    paused = data.paused;
                    setUsersPausedRealInfo(data.realInfo);
                }
                if (!data.isMatching && !paused) {
                    doc?.getElementById('authtics-dev-panel-root')?.remove();
                }
                const resp2 = await fetch('/__authtics_host_apis__/get-dat-restriction-missing', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({})
                });
                if (resp2.status === 200) {
                    const data = await resp2.json();
                    setDatRestrictionMissing(!data.enabled);
                    datRestrictionMissingRef.current = !data.enabled;
                    return "";
                }
            }
            return "";
        };
        const isSSR = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-is-ssr', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setIsSsr(data.isSsr);
            }
        };
        const getUsage = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-cpu_memory', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                cpuUsage.current = data.cpuUsage;
                heapUsage.current = data.heapUsed;
            }
        };
        const getHostVersion = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-host-ver', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setHostVersion(data.version);
            }
        };
        const getCompilerType = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-compiler', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setCompilerType(data.compiler);
            }
        };
        const getHttpVersion = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-connection', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setHttpVersion(data.httpVersion);
            }
        };
        const formatSeconds = (totalSeconds) => {
            const days = Math.floor(totalSeconds / 86400);
            const hours = Math.floor(totalSeconds / 3600);
            const minutes = Math.floor((totalSeconds % 3600) / 60);
            const seconds = totalSeconds % 60;
            const pad = (num) => String(num).padStart(2, '0');
            return `${pad(days)}:${pad(hours)}:${pad(minutes)}:${pad(seconds)}`;
        };
        const userSettings = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-user-settings', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setErrorBoxOpenable(data.showError);
            }
        };
        const update = async () => {
            const resp = await fetch('/__authtics_host_apis__/get-server-info', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({})
            });
            if (resp.status === 200) {
                const data = await resp.json();
                setUptime(formatSeconds(data.uptime));
                setRequests(data.totalRequests);
                setUsers(data.users);
            }
        };
        setTimeout(async () => {
            await getDatMatching();
            setTimeout(() => {
                if (datRestrictionMissingRef.current === false) {
                    getPath();
                    getCompilerType();
                    getHttpVersion();
                    isSSR();
                    userSettings();
                    getHostVersion();
                    setInterval(() => {
                        update();
                        getUsage();
                    }, 500);
                }
            }, 250);
        }, 250);
    }, []);
    const restart = async () => {
        await fetch('/__authtics_host_apis__/restart', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({})
        });
    };
    const shutDown = async () => {
        await fetch('/__authtics_host_apis__/shutdown', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({})
        });
    };
    useEffect(() => {
        const animationId = requestAnimationFrame(loop);
        const intervalId = setInterval(() => {
            if (fpsSpanRef.current) {
                fpsSpanRef.current.textContent = frameCount.current.toString();
                frames.current = frameCount.current;
            }
            frameCount.current = 0;
        }, 1000);
        return () => {
            cancelAnimationFrame(animationId);
            clearInterval(intervalId);
        };
    }, []);
    const pendingSent = useRef(0);
    const pendingReceived = useRef(0);
    const [sentHistory, setSentHistory] = useState(Array(MAX_POINTS).fill(0));
    const [receivedHistory, setReceivedHistory] = useState(Array(MAX_POINTS).fill(0));
    const [cpuUsageHistory, setCpuUsageHistory] = useState(Array(MAX_POINTS).fill(0));
    const [heapUsageHistory, setHeapUsageHistory] = useState(Array(MAX_POINTS).fill(0));
    const drawerHeight = !open
        ? '0px'
        : consoleOpen || networkOpen || routeOpen || webpageInfoOpen || clientInfoOpen || aboutHostOpen
            ? '320px'
            : '320px';
    useEffect(() => {
        const interval = setInterval(async () => {
            try {
                const timeStart = Date.now();
                const resp = await fetch('/__authtics_host_apis__/ping', {
                    method: 'POST',
                    headers: { 'Content-Type': 'application/json' },
                    body: JSON.stringify({})
                });
                if (resp.status === 200) {
                    const data = await resp.json();
                    if (data.status === 'ok')
                        setPing__(Date.now() - timeStart);
                }
            }
            catch { }
        }, 1000);
        return () => clearInterval(interval);
    }, []);
    useEffect(() => {
        const interval = setInterval(() => {
            const s = pendingSent.current;
            const r = pendingReceived.current;
            const c = cpuUsage.current;
            const h = heapUsage.current;
            pendingSent.current = 0;
            pendingReceived.current = 0;
            cpuUsage.current = 0;
            heapUsage.current = 0;
            setCpuUsageHistory(prev => [...prev.slice(-(MAX_POINTS - 1)), c]);
            setHeapUsageHistory(prev => [...prev.slice(-(MAX_POINTS - 1)), h]);
            setSentHistory(prev => [...prev.slice(-(MAX_POINTS - 1)), s]);
            setReceivedHistory(prev => [...prev.slice(-(MAX_POINTS - 1)), r]);
        }, 1000);
        return () => clearInterval(interval);
    }, []);
    useEffect(() => {
        installConsoleInterceptor();
        installNetworkInterceptor((f) => setNetworkFetches(prev => [...prev.slice(-49), f]), (f) => setNetworkFetches(prev => prev.map(p => p.id === f.id ? f : p)), (f) => setEventSources(prev => [...prev.slice(-49), f]), (f) => setEventSources(prev => prev.map(p => p.id === f.id ? f : p)), (f) => setWebSockets(prev => [...prev.slice(-49), f]), (f) => setWebSockets(prev => prev.map(p => p.id === f.id ? f : p)), (f) => setResourcesFound(prev => [...prev.slice(-49), f]), (f) => setResourcesFound(prev => prev.map(p => p.id === f.id ? f : p)), (f) => setResourcesFound(prev => prev.filter(p => p.id !== f.id)), (sent) => {
            pendingSent.current += sent;
        }, (received) => {
            pendingReceived.current += received;
        });
        const handler = (e) => {
            const detail = e.detail;
            setLogs(prev => [...prev.slice(-99), { ...detail, id: Date.now() + Math.random() }]);
            if (detail.type === 'Error') {
                setHasIssue(true);
                setErrorMsgs(prev => [...prev, detail.message]);
                setErrorBoxOpen(true);
            }
            ;
        };
        try {
            const existingToken = cookie.get('dat');
            if (existingToken) {
                setToken(existingToken);
                let length_ = existingToken?.split('');
                let length;
                length = length_?.length;
                setTokenLength(length);
            }
            else {
                const newToken = generateToken();
                try {
                    let localStorageToken;
                    try {
                        localStorageToken = localStorage.getItem('dat');
                    }
                    catch (e) {
                    }
                    cookie.set('dat', localStorageToken || newToken);
                    setToken(newToken);
                    let length_ = newToken?.split('');
                    let length;
                    length = length_?.length;
                    setTokenLength(length);
                }
                catch (error) {
                    console.error('[Authtics Host] Error storing token in localStorage:', error);
                }
            }
        }
        catch (error) {
        }
        window.addEventListener('authtics:log', handler);
        setInterval(() => setStatus(window.__status_code__ || 200), 1000);
        setPathName(window.location.pathname);
        setHostName(window.location.hostname);
        setDoc(document);
        setProtocol(window.location.protocol.replace(':', '').toUpperCase());
        return () => window.removeEventListener('authtics:log', handler);
    }, []);
    const getOS = () => {
        const ua = window.navigator.userAgent;
        if (/Android/i.test(ua))
            return "Android";
        if (/Linux/i.test(ua)) {
            return (navigator.maxTouchPoints > 0) ? "Android (Desktop Mode)" : "Linux";
        }
        if (/iPhone|iPad|iPod/i.test(ua))
            return "iOS";
        if (/Mac/i.test(ua))
            return "MacOS";
        if (/Win/i.test(ua))
            return "Windows";
        return "Unknown";
    };
    useEffect(() => {
        setHasIssue(errorCount > 0);
    }, [errorCount, warnCount]);
    const toggleFilter = (type) => {
        setFilters(prev => ({ ...prev, [type]: !prev[type] }));
    };
    const errorBoxRef = useRef(null);
    setInterval(() => {
        if (errorBoxRef.current) {
            if (errorBoxOpen && errorBoxRef.current.style.opacity == '0') {
                errorBoxRef.current.style.opacity = '1';
            }
            else if (!errorBoxOpen && errorBoxRef.current.style.opacity == '1') {
                errorBoxRef.current.style.opacity = '0';
            }
        }
    }, 500);
    useEffect(() => {
        const link = document.createElement('link');
        link.rel = 'stylesheet';
        link.href = 'https://cdnjs.cloudflare.com/ajax/libs/prism/1.29.0/themes/prism-okaidia.min.css';
        document.head.appendChild(link);
        const style = document.createElement('style');
        style.textContent = `.token.punctuation  { color: #333333 !important; }
.token.operator     { color: #d73a49 !important; }
.token.variable     { color: #001080 !important; }
.token.entity       { color: #d73a49 !important; }
.token.url          { color: #0070c1 !important; }
.token.string       { color: #a31515 !important; }
.token.function     { color: #795e26 !important; }
.token.keyword      { color: #af00db !important; }
.token.number       { color: #098658 !important; }
.token.boolean      { color: #0000ff !important; }
.token.comment      { color: #6a9955 !important; }
.token.class-name   { color: #267f99 !important; }
.token.property     { color: #001080 !important; }
.token.builtin      { color: #0070c1 !important; }
.token.template-string        { color: #a31515 !important; }
.token.template-punctuation   { color: #d73a49 !important; }
.token.interpolation          { color: #333333 !important; }
.token.interpolation-punctuation { color: #d73a49 !important; }`;
        document.head.appendChild(style);
    }, []);
    useEffect(() => {
        const __ = async () => {
            await fetch(window.location.pathname, {
                method: 'GET'
            });
        };
        __();
    }, []);
    const pause_Resume_Users = async () => {
        if (usersPausedRealInfo) {
            const resp = await fetch('/__authtics_host_apis__/set-users-paused', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    paused: false
                })
            });
            if (resp.status === 200) {
                setUsersPausedRealInfo(false);
            }
        }
        else {
            const resp = await fetch('/__authtics_host_apis__/set-users-paused', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    paused: true
                })
            });
            if (resp.status === 200) {
                setUsersPausedRealInfo(true);
            }
        }
    };
    return (_jsxs(_Fragment, { children: [_jsx("div", { style: {
                    position: 'absolute',
                    zIndex: 999999,
                    background: 'rgba(0,0,0,0.5)',
                    top: 0,
                    left: 0,
                    width: '100%',
                    height: '100%',
                    display: errorBoxOpen && errorBoxOpenable && errorMsgs.length > 0 ? 'flex' : 'none',
                    justifyContent: 'center',
                    alignItems: 'center',
                    transition: errorBoxOpen && errorBoxOpenable && errorMsgs.length > 0 ? 'opacity 0.25s ease' : '',
                }, ref: errorBoxRef, children: errorBoxOpen && errorBoxOpenable && errorMsgs.length > 0 ? (_jsxs("div", { style: {
                        background: '#f7f7f7',
                        border: '1px solid #e6e3e3',
                        borderRadius: '10px',
                        padding: '5px',
                        boxShadow: '0 8px 32px rgba(0,0,0,0.15)',
                        display: 'flex',
                        flexDirection: 'column',
                        zIndex: 9999999,
                        width: '320px'
                    }, children: [_jsxs("div", { style: {
                                padding: '6px 8px',
                                display: 'flex',
                                alignItems: 'center',
                                gap: '6px',
                                fontFamily: 'system-ui, sans-serif',
                            }, children: [_jsxs("div", { style: {
                                        background: 'rgba(0,0,0,0.15)',
                                        borderRadius: '100px',
                                        padding: '4px 8px',
                                        fontSize: '12px',
                                        display: 'flex',
                                        justifyContent: 'center',
                                        alignItems: 'center',
                                        whiteSpace: 'nowrap',
                                        gap: '4px'
                                    }, children: [_jsx(ShieldCheck, { size: 16, color: "#f472b6", style: { flexShrink: 0 } }), " Authtics Host"] }), _jsxs("p", { style: { color: 'red', fontSize: '12px' }, children: ["Error found (", selectedError, "/", errorMsgs.length, ")"] }), _jsx(PanelButton, { onClick: () => { setErrorBoxOpen(false); }, style: { marginLeft: 'auto', }, children: "Close" })] }), _jsx("div", { style: {
                                padding: '6px 2px',
                                display: 'flex',
                                alignItems: 'center',
                                gap: '6px',
                                fontFamily: 'system-ui, sans-serif'
                            }, children: _jsxs("p", { style: { color: 'red', fontSize: '24px', gap: '6px', display: 'flex', justifyContent: 'center', alignItems: 'center', marginLeft: '10px' }, children: [_jsx(CircleX, { size: 24, color: 'red' }), errorMsgs[selectedError - 1]] }) }), _jsxs("div", { style: {
                                padding: '6px 8px',
                                display: 'flex',
                                alignItems: 'center',
                                gap: '6px',
                                fontFamily: 'system-ui, sans-serif'
                            }, children: [_jsxs(PanelButton, { onClick: () => { setSelectedError(selectedError - 1 < 1 ? selectedError : selectedError - 1); }, style: { marginLeft: 'auto', width: '50%', justifyContent: 'space-between', display: 'flex' }, children: ["Previous ", _jsx(ChevronLeft, { size: 12 })] }), _jsxs(PanelButton, { onClick: () => { setSelectedError(selectedError + 1 > errorMsgs.length ? selectedError : selectedError + 1); }, style: { marginLeft: 'auto', width: '50%', justifyContent: 'space-between', display: 'flex', background: '#34a8eb' }, children: ["Next ", _jsx(ChevronRight, { size: 12 })] })] })] })) : (_jsx(_Fragment, {})) }), _jsxs("div", { style: {
                    position: 'fixed',
                    bottom: '15px',
                    right: '15px',
                    zIndex: 999999,
                    fontFamily: 'system-ui, sans-serif',
                    fontSize: '13px',
                    display: 'flex',
                    flexDirection: 'column',
                    alignItems: 'flex-end',
                    gap: '8px',
                }, children: [_jsx("div", { style: {
                            maxHeight: drawerHeight,
                            opacity: open ? 1 : 0,
                            pointerEvents: open ? 'auto' : 'none',
                            overflow: (!consoleOpen && !networkOpen && !aboutHostOpen && !webpageInfoOpen && !clientInfoOpen ? 'hidden auto' : 'hidden'),
                            borderRadius: '10px',
                            transition: 'max-height 0.35s ease, opacity 0.25s ease',
                        }, children: _jsx("div", { style: {
                                background: '#f7f7f7',
                                border: '1px solid #e6e3e3',
                                borderRadius: '10px',
                                width: '320px',
                                display: 'flex',
                                flexDirection: 'column',
                                boxShadow: '0 8px 32px rgba(0,0,0,0.15)',
                                boxSizing: 'border-box',
                                overflow: 'hidden',
                            }, children: consoleOpen ? (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderBottom: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            alignItems: 'center',
                                            gap: '6px',
                                        }, children: [_jsxs(PanelButton, { onClick: () => setConsoleOpen(false), style: { color: '#555' }, children: [_jsx(ArrowLeft, { size: 12 }), " Back"] }), _jsxs("span", { style: { fontSize: '12px', display: 'flex', alignItems: 'center', gap: '4px', color: '#333' }, children: [_jsx(Terminal, { size: 12, color: "#16a34a" }), " Console"] }), _jsxs("span", { style: { fontSize: '10px', color: '#aaa', marginLeft: '2px' }, children: [filteredLogs.length, " message", filteredLogs.length !== 1 ? 's' : '', " shown, ", logs.length, " in total"] }), _jsx(PanelButton, { onClick: () => setLogs([]), style: { color: '#aaa', marginLeft: 'auto' }, children: "Clear" })] }), _jsx("div", { style: {
                                            display: 'flex',
                                            gap: '4px',
                                            padding: '5px 8px',
                                            borderBottom: '1px solid #eee',
                                            background: '#fafafa',
                                            flexWrap: 'wrap',
                                        }, children: [...LOG_TYPES].map(type => (_jsxs("label", { style: {
                                                display: 'flex',
                                                alignItems: 'center',
                                                gap: '3px',
                                                fontSize: '10px',
                                                color: filters[type] ? FILTER_COLORS[type] : '#bbb',
                                                cursor: 'pointer',
                                                padding: '2px 6px',
                                                borderRadius: '10px',
                                                background: filters[type] ? `${FILTER_COLORS[type]}18` : 'transparent',
                                                border: `1px solid ${filters[type] ? FILTER_COLORS[type] + '44' : '#e0e0e0'}`,
                                                transition: 'all 0.15s ease',
                                                userSelect: 'none',
                                            }, children: [_jsx("input", { type: "checkbox", checked: filters[type], onChange: () => toggleFilter(type), style: { display: 'none' } }), type] }, type))) }), _jsx("div", { style: { overflowY: 'auto', maxHeight: '220px' }, ref: consoleScrollRef, onScroll: handleScroll, children: filteredLogs.length === 0
                                            ? _jsx("div", { style: { padding: '16px 12px', color: '#aaa', fontSize: '12px' }, children: "No logs yet." })
                                            : [...filteredLogs].map(log => {
                                                const s = TYPE_STYLE[log.type];
                                                return (_jsxs("div", { style: {
                                                        display: 'flex',
                                                        gap: '8px',
                                                        padding: '5px 10px',
                                                        background: s.bg,
                                                        borderBottom: '1px solid #f0f0f0',
                                                        alignItems: 'flex-start',
                                                    }, children: [s.icon && (_jsx("span", { style: { color: s.color, marginTop: '1px', flexShrink: 0, display: 'flex' }, children: s.icon })), _jsx("span", { style: {
                                                                color: s.textColor ?? '#555',
                                                                fontSize: '11px',
                                                                fontFamily: 'monospace',
                                                                wordBreak: 'break-all',
                                                            }, children: log.message })] }, log.id));
                                            }) })] })) : networkOpen ? (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderBottom: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            alignItems: 'center',
                                            gap: '6px',
                                        }, children: [_jsxs(PanelButton, { onClick: () => setNetworkOpen(false), style: { color: '#555' }, children: [_jsx(ArrowLeft, { size: 12 }), " Back"] }), _jsxs("span", { style: { fontSize: '12px', display: 'flex', alignItems: 'center', gap: '4px', color: '#333' }, children: [_jsx(Wifi, { size: 12, color: "#3b82f6" }), " Network"] })] }), _jsxs("div", { style: { overflowY: 'auto', maxHeight: '260px', padding: '10px', display: 'flex', flexDirection: 'column', gap: '8px' }, children: [_jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { children: "Ping" }), _jsx("span", { style: { color: ping__ < 100 ? '#16a34a' : ping__ < 300 ? '#d97706' : '#e53e3e', fontFamily: 'monospace' }, children: ping__ > 0 ? `${ping__}ms` : '??ms' })] }), datRestrictionMissing ? (_jsxs("span", { style: { display: 'flex', opacity: 0.5, fontSize: 10, justifyContent: 'center', alignItems: 'center', gap: '4px' }, children: [_jsx(Info, { size: 10 }), "this analytic is unavaliable. Set \"dat\" in the Authtics Host config to set restriction and then restart."] })) : (_jsx(_Fragment, {})), _jsx(Sparkline, { data: receivedHistory, color: "#3b82f6", label: "Received", unit: " bytes" }), _jsx(Sparkline, { data: sentHistory, color: "#16a34a", label: "Sent", unit: " bytes" }), _jsxs("span", { style: { fontSize: '15px', color: '#333', fontWeight: 'bold' }, children: ["Fetches \u2014 ", networkFetches.length, " total fetches"] }), networkFetches.length === 0
                                                ? _jsx("div", { style: { color: '#aaa', fontSize: '12px', padding: '4px 2px' }, children: "No network requests yet." })
                                                : [...networkFetches].reverse().map((f, i) => (_jsxs("div", { style: {
                                                        background: '#fff',
                                                        display: 'flex',
                                                        flexDirection: 'column',
                                                        borderRadius: '10px',
                                                        border: '1px solid #eee',
                                                        fontSize: '11px',
                                                        fontFamily: 'monospace',
                                                    }, children: [_jsxs("div", { style: { display: 'flex',
                                                                justifyContent: 'space-between', gap: '8px', padding: '6px 10px', borderBottom: '1px solid #eee' }, children: [_jsx("span", { style: { color: '#3b82f6' }, children: f.method }), _jsx("span", { style: { flex: 1, overflow: 'hidden', textOverflow: 'ellipsis', whiteSpace: 'nowrap', color: '#333' }, children: f.url }), _jsx("span", { style: { color: f.status ? (f.status < 400 ? '#16a34a' : '#e53e3e') : 'orange' }, children: f.status || "Awaiting for response..." })] }), f.method === "POST" ? (_jsxs("div", { style: { padding: '10px 10px', display: 'flex', flexDirection: 'column', gap: '8px' }, children: [_jsx("p", { style: { color: 'orange', fontFamily: 'system-ui, sans-serif', margin: 0 }, children: "Sent: " }), _jsx("div", { children: f.sent })] })) : (_jsx(_Fragment, {})), _jsxs("div", { style: { padding: '10px 10px', display: 'flex', flexDirection: 'column', gap: '8px' }, children: [_jsx("p", { style: { color: 'green', fontFamily: 'system-ui, sans-serif', margin: 0 }, children: "Received: " }), _jsx(TruncatedText, { content: f.response || "...", id: `${f.id}_fetch` })] }), f.response ? (f.response.split('').length > 100 ? (_jsxs(PanelButton, { onClick2: (e) => { const obj___ = document.getElementById(`${f.id}_fetch`); if (obj___ && !obj___.hasAttribute('expanded')) {
                                                                obj___.innerText = f.response;
                                                                obj___.setAttribute('expanded', '');
                                                                e.currentTarget.innerText = 'View less';
                                                            }
                                                            else if (obj___ && obj___.hasAttribute('expanded')) {
                                                                obj___.innerText = f.response?.slice(0, 100) + "...";
                                                                obj___.removeAttribute('expanded');
                                                                e.currentTarget.innerText = `View more (${f.response ? (f.response.split('').length - 100) : 0} bytes hidden)`;
                                                            } }, style: { padding: '8px 10px', borderRadius: '10px', background: '#34a8eb', color: '#fff', border: 'none', cursor: 'pointer', margin: '10px', textAlign: 'center', fontFamily: 'system-ui, sans-serif' }, children: ["View more (", f.response ? (f.response.split('').length - 100) : 0, " bytes hidden)"] })) : (_jsx(_Fragment, {}))) : (_jsx(_Fragment, {}))] }, i))), _jsx("span", { style: { fontSize: '15px', color: '#333', fontWeight: 'bold' }, children: "Active connections" }), eventSources.length === 0 ? (_jsx("div", { style: { color: '#aaa', fontSize: '12px', padding: '4px 2px' }, children: "No active connections." })) :
                                                [...eventSources, ...webSockets].reverse().map((es, i) => (_jsx("div", { style: {
                                                        background: '#fff',
                                                        display: 'flex',
                                                        flexDirection: 'column',
                                                        borderRadius: '10px',
                                                        border: '1px solid #eee',
                                                        fontSize: '11px',
                                                        fontFamily: 'monospace'
                                                    }, children: _jsxs("div", { style: { display: 'flex', justifyContent: 'space-between', gap: '8px', padding: '6px 10px' }, children: [_jsxs("span", { style: { color: '#3b82f6', display: 'flex', gap: '4px', justifyContent: 'center', alignItems: 'center' }, children: [es.type === "EventSource" ? _jsx(Zap, { color: "green", size: 12 }) : es.type === "WebSocket" ? _jsx(Globe, { color: "green", size: 12 }) : _jsx(_Fragment, {}), es.type === "EventSource" ? "Event Source" : es.type === "WebSocket" ? "Web Socket" : "???"] }), _jsx("span", { style: { flex: 1, overflow: 'hidden', textOverflow: 'ellipsis', whiteSpace: 'nowrap', color: '#333' }, children: es.url }), _jsx("span", { style: { color: es.status === 'Connected' ? '#16a34a' : es.status === 'Disconnected' ? '#e53e3e' : 'orange' }, children: es.status === 'Connected' ? 'Active' : es.status === 'Disconnected' ? 'Disconnected' : es.status === 'Connecting...' ? '...' : es.status === 'Connection Failed' ? 'Failed' : 'Unknown' })] }) }, i))), _jsxs("span", { style: { fontSize: '15px', color: '#333', fontWeight: 'bold' }, children: ["Resources Found \u2014 ", resourcesFound.length, " total resources"] }), resourcesFound.length === 0 ? (_jsx("div", { style: { color: '#aaa', fontSize: '12px', padding: '4px 2px' }, children: "No resources found." })) :
                                                [...resourcesFound].reverse().map((r, i) => (_jsx("div", { style: {
                                                        background: '#fff',
                                                        display: 'flex',
                                                        flexDirection: 'column',
                                                        borderRadius: '10px',
                                                        border: '1px solid #eee',
                                                        fontSize: '11px',
                                                        fontFamily: 'monospace'
                                                    }, children: _jsxs("div", { style: { display: 'flex', justifyContent: 'space-between', gap: '8px', padding: '6px 10px' }, children: [_jsx("span", { style: { color: '#3b82f6', display: 'flex', gap: '4px', justifyContent: 'center', alignItems: 'center' }, children: r.type }), _jsx("span", { style: { flex: 1, color: '#333' }, children: r.url })] }) }, i)))] })] })) : routeOpen ? (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderBottom: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            alignItems: 'center',
                                            gap: '6px',
                                        }, children: [_jsxs(PanelButton, { onClick: () => setRouteOpen(false), style: { color: '#555' }, children: [_jsx(ArrowLeft, { size: 12 }), " Back"] }), _jsxs("span", { style: { fontSize: '12px', display: 'flex', alignItems: 'center', gap: '4px', color: '#333' }, children: [_jsx(Route, { size: 12, color: "#3b82f6" }), " Route"] }), _jsx("span", { style: { fontSize: '10px', color: '#aaa', marginLeft: '2px' }, children: pathName })] }), datRestrictionMissing ? (_jsx(SecuredTab, {})) : (_jsx(_Fragment, {})), _jsxs("div", { style: { overflowY: 'auto', maxHeight: '260px', padding: '10px', display: (datRestrictionMissing ? 'none' : 'flex'), flexDirection: 'column', gap: '8px' }, children: [_jsx("span", { style: { fontSize: '24px', color: '#333', fontWeight: 'bold' }, children: pathName }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Load Status Code" }), _jsx("span", { style: { color: status === 200 ? '#16a34a' : '#e53e3e', fontFamily: 'monospace' }, children: status })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "File Path" }), _jsx("span", { style: { color: 'blue', fontFamily: 'monospace' }, children: path })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "File Name" }), _jsx("span", { style: { color: 'blue', fontFamily: 'monospace' }, children: path?.split('/').pop() })] })] })] })) : webpageInfoOpen ? (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderBottom: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            alignItems: 'center',
                                            gap: '6px',
                                        }, children: [_jsxs(PanelButton, { onClick: () => setWebpageInfoOpen(false), style: { color: '#555' }, children: [_jsx(ArrowLeft, { size: 12 }), " Back"] }), _jsxs("span", { style: { fontSize: '12px', display: 'flex', alignItems: 'center', gap: '4px', color: '#333' }, children: [_jsx(Globe, { size: 12, color: "#3b82f6" }), " Webpage Info"] })] }), datRestrictionMissing ? (_jsx(SecuredTab, {})) : (_jsx(_Fragment, {})), _jsxs("div", { style: { overflowY: 'auto', maxHeight: '260px', padding: '10px', display: (datRestrictionMissing ? 'none' : 'flex'), flexDirection: 'column', gap: '8px' }, children: [_jsxs("span", { style: { fontSize: '12px', color: '#333' }, children: [_jsx("span", { style: { fontSize: '24px', color: '#333', fontWeight: 'bold' }, children: pathName }), " on ", _jsx("span", { style: { fontSize: '12px', color: '#333', fontWeight: 'bold' }, children: hostName })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Protocol" }), _jsx("span", { style: { fontSize: '12px', color: 'orange' }, children: protocol })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "SSL" }), _jsx("span", { style: { fontSize: '12px', color: (protocol === 'HTTPS' ? '#16a34a' : '#e53e3e') }, children: protocol === 'HTTPS' ? 'Yes' : 'No' })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Renderer Type" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: isSsr ? 'SSR (Server-Side Rendering)' : 'CSR (Client-Side Rendering)' })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Compiler Type" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: compilerType })] })] })] })) : clientInfoOpen ? (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderBottom: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            alignItems: 'center',
                                            gap: '6px',
                                        }, children: [_jsxs(PanelButton, { onClick: () => setClientInfoOpen(false), style: { color: '#555' }, children: [_jsx(ArrowLeft, { size: 12 }), " Back"] }), _jsxs("span", { style: { fontSize: '12px', display: 'flex', alignItems: 'center', gap: '4px', color: '#333' }, children: [_jsx(Info, { size: 12, color: "#3b82f6" }), " Client Info"] })] }), datRestrictionMissing ? (_jsx(SecuredTab, {})) : (_jsx(_Fragment, {})), _jsxs("div", { style: { overflowY: 'auto', maxHeight: '260px', padding: '10px', display: (datRestrictionMissing ? 'none' : 'flex'), flexDirection: 'column', gap: '8px' }, children: [_jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "HTTP Version" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: "v" + httpVersion })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "User Agent" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: navigator.userAgent })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "OS" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: getOS() })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "FPS" }), _jsx("span", { style: { fontSize: '12px', color: (frames.current < 30 ? '#e53e3e' : frames.current < 60 ? '#d97706' : '#16a34a') }, ref: fpsSpanRef, children: "\u2014" })] })] })] })) : aboutHostOpen ? (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderBottom: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            alignItems: 'center',
                                            gap: '6px',
                                        }, children: [_jsxs(PanelButton, { onClick: () => setAboutHostOpen(false), style: { color: '#555' }, children: [_jsx(ArrowLeft, { size: 12 }), " Back"] }), _jsxs("span", { style: { fontSize: '12px', display: 'flex', alignItems: 'center', gap: '4px', color: '#333' }, children: [_jsx(ShieldCheck, { size: 12, color: "#f472b6" }), " About Host"] })] }), datRestrictionMissing ? (_jsx(SecuredTab, {})) : (_jsx(_Fragment, {})), _jsxs("div", { style: { overflowY: 'auto', maxHeight: '260px', padding: '10px', display: (datRestrictionMissing ? 'none' : 'flex'), flexDirection: 'column', gap: '8px' }, children: [_jsxs("span", { style: { fontSize: '24px', color: '#333', fontWeight: 'bold', textAlign: 'center', display: 'flex', alignItems: 'center', flexDirection: 'column', gap: '6px' }, children: [_jsx(ShieldCheck, { size: 36, color: "#f472b6" }), "Authtics Host"] }), _jsx("span", { style: { fontSize: '12px', color: '#333', textAlign: 'center' }, children: "A lightweight, fast, and secure web server for modern web applications." }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Version" }), _jsxs("span", { style: { fontSize: '12px', color: 'blue' }, children: ["v", hostVersion] })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Made By" }), _jsx("span", { style: { fontSize: '12px', color: 'orange' }, children: "bananakitssu" })] }), _jsxs("a", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                    cursor: 'pointer',
                                                    textDecoration: 'none'
                                                }, href: "https://www.npmjs.com/package/@bananacool467/authtics-host", children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Documentation" }), _jsx("span", { style: { fontSize: '12px', color: '#333' }, children: _jsx(ChevronRight, { size: 12 }) })] })] })] })) : serverInfoOpen ? (_jsxs(_Fragment, { children: [_jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderBottom: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            alignItems: 'center',
                                            gap: '6px',
                                        }, children: [_jsxs(PanelButton, { onClick: () => setServerInfoOpen(false), style: { color: '#555' }, children: [_jsx(ArrowLeft, { size: 12 }), " Back"] }), _jsxs("span", { style: { fontSize: '12px', display: 'flex', alignItems: 'center', gap: '4px', color: '#333' }, children: [_jsx(Server, { size: 12, color: "#f472b6" }), " Server Info"] })] }), datRestrictionMissing ? (_jsx(SecuredTab, {})) : (_jsx(_Fragment, {})), _jsxs("div", { style: { overflowY: 'auto', maxHeight: '260px', padding: '10px', display: (datRestrictionMissing ? 'none' : 'flex'), flexDirection: 'column', gap: '8px' }, children: [_jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Server uptime" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: uptime })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Total requests" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: requests })] }), _jsxs("div", { style: {
                                                    display: 'flex',
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333' }, children: "Total current users" }), _jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: users })] }), _jsx(Sparkline, { data: cpuUsageHistory, color: "#3b82f6", label: "CPU Usage", unit: "" }), _jsx(Sparkline, { data: heapUsageHistory, color: "#3b82f6", label: "Heap Used", unit: "" })] })] })) : (_jsxs(_Fragment, { children: [_jsxs("div", { style: { padding: '8px', gap: '4px', display: 'flex', flexDirection: 'column' }, children: [_jsxs(PanelButton, { onClick: () => setConsoleOpen(true), fullWidth: true, style: { color: '#333', justifyContent: 'space-between', height: '32px' }, children: [_jsxs("span", { style: { display: 'flex', alignItems: 'center', gap: '6px' }, children: [_jsx(Terminal, { size: 12, color: "#16a34a" }), " Console"] }), logs.length > 0 && (_jsx("span", { style: { color: '#111', borderRadius: '100px', padding: '1px 6px', fontSize: '10px' }, children: logs.length }))] }), _jsxs(PanelButton, { onClick: () => setRouteOpen(true), fullWidth: true, style: { color: '#333', justifyContent: 'space-between', height: '32px' }, children: [_jsxs("span", { style: { display: 'flex', alignItems: 'center', gap: '6px' }, children: [_jsx(Route, { size: 12, color: "#3b82f6" }), " Route"] }), _jsxs("span", { style: { color: '#111', borderRadius: '100px', padding: '1px 6px', fontSize: '10px', display: 'flex', alignItems: 'center', justifyContent: 'center', gap: '4px' }, children: [datRestrictionMissing && _jsx(TriangleAlert, { size: 12, color: "orange" }), pathName] })] }), _jsxs(PanelButton, { onClick: () => setWebpageInfoOpen(true), fullWidth: true, style: { color: '#333', justifyContent: 'space-between', height: '32px' }, children: [_jsxs("span", { style: { display: 'flex', alignItems: 'center', gap: '6px' }, children: [_jsx(Globe, { size: 12, color: "#3b82f6" }), " Webpage Info"] }), datRestrictionMissing && _jsx(TriangleAlert, { size: 12, color: "orange" })] }), _jsx(PanelButton, { onClick: () => setNetworkOpen(true), fullWidth: true, style: { color: '#333', justifyContent: 'space-between', height: '32px' }, children: _jsxs("span", { style: { display: 'flex', alignItems: 'center', gap: '6px' }, children: [_jsx(Wifi, { size: 12, color: "#3b82f6" }), " Network"] }) }), _jsxs(PanelButton, { onClick: () => setClientInfoOpen(true), fullWidth: true, style: { color: '#333', justifyContent: 'space-between', height: '32px' }, children: [_jsxs("span", { style: { display: 'flex', alignItems: 'center', gap: '6px' }, children: [_jsx(Info, { size: 12, color: "#3b82f6" }), " Client Info"] }), datRestrictionMissing && _jsx(TriangleAlert, { size: 12, color: "orange" })] }), _jsxs(PanelButton, { onClick: () => setAboutHostOpen(true), fullWidth: true, style: { color: '#333', justifyContent: 'space-between', height: '32px' }, children: [_jsxs("span", { style: { display: 'flex', alignItems: 'center', gap: '6px' }, children: [_jsx(ShieldCheck, { size: 12, color: "#f472b6" }), " About Host"] }), datRestrictionMissing && _jsx(TriangleAlert, { size: 12, color: "orange" })] }), _jsxs(PanelButton, { onClick: () => setServerInfoOpen(true), fullWidth: true, style: { color: '#333', justifyContent: 'space-between', height: '32px' }, children: [_jsxs("span", { style: { display: 'flex', alignItems: 'center', gap: '6px' }, children: [_jsx(Server, { size: 12, color: "#f472b6" }), " Server Info"] }), datRestrictionMissing && _jsx(TriangleAlert, { size: 12, color: "orange" })] })] }), _jsxs("div", { style: {
                                            padding: '6px 8px',
                                            borderTop: '1px solid #ddd',
                                            background: '#fff',
                                            display: 'flex',
                                            gap: '6px',
                                            overflowY: 'auto',
                                            maxHeight: '260px',
                                            flexDirection: 'column',
                                        }, children: [_jsxs("div", { style: {
                                                    justifyContent: 'space-between',
                                                    padding: '6px 10px',
                                                    background: '#fff',
                                                    borderRadius: '10px',
                                                    border: '1px solid #eee',
                                                    fontSize: '12px',
                                                    color: '#333',
                                                    display: 'flex'
                                                }, children: [_jsx("span", { style: { fontSize: '12px', color: '#333', position: 'relative' }, children: "DAT (Developer Access Token)" }), _jsxs("div", { style: { position: 'relative',
                                                            display: 'inline-flex' }, children: [_jsx("span", { style: { fontSize: '12px', color: 'blue' }, children: token }), tokenHidden && (_jsxs("div", { style: {
                                                                    position: 'absolute',
                                                                    top: 0, left: 0,
                                                                    marginTop: '-2.5px',
                                                                    width: '100%', height: '100%',
                                                                    backdropFilter: 'blur(10px)',
                                                                    backgroundColor: 'rgba(255,255,255,0.3)',
                                                                    cursor: 'pointer',
                                                                    padding: '2.5px 2.5px',
                                                                    borderRadius: '8px',
                                                                    textAlign: 'center',
                                                                    display: 'flex',
                                                                    justifyContent: 'center',
                                                                    alignItems: 'center',
                                                                    gap: '4px'
                                                                }, onClick: () => setTokenHidden(false), children: [_jsx(Eye, { size: 12 }), "Click to reveal"] }))] })] }), tokenLength !== 16 ? (_jsx(_Fragment, { children: _jsxs("div", { style: {
                                                        display: 'flex',
                                                        background: '#ba4e4e',
                                                        borderRadius: '10px',
                                                        padding: '6px 10px',
                                                        justifyContent: 'left',
                                                        alignItems: 'center',
                                                        gap: '6px',
                                                        color: '#fff',
                                                        flexDirection: 'row',
                                                    }, children: [_jsx(CircleX, { size: 12 }), "Invalid Developer Access Token (DAT)"] }) })) : (_jsx(_Fragment, {})), hostVersion === "0.0.1-beta.9" ? (_jsx(_Fragment, { children: _jsx("div", { style: {
                                                        display: 'flex',
                                                        background: 'orange',
                                                        borderRadius: '10px',
                                                        padding: '6px 10px',
                                                        justifyContent: 'left',
                                                        alignItems: 'center',
                                                        gap: '6px',
                                                        color: '#fff',
                                                        flexDirection: 'row',
                                                    }, children: "Updates to security. Apps with version 0.0.1-beta.9 and over will now require DAT restriction to access remote features like restarting and shutting down the app." }) })) : (_jsx(_Fragment, {})), _jsx(PanelButton, { onClick: () => { restart(); }, fullWidth: true, style: { color: '#fff', justifyContent: 'center', height: '32px', background: '#ba4e4e' }, disabled: datRestrictionMissing, children: _jsx("span", { style: { display: 'flex', alignItems: 'center', gap: '6px', textAlign: 'center' }, children: "Restart App" }) }), datRestrictionMissing ? (_jsxs("span", { style: { display: 'flex', opacity: 0.5, fontSize: 10, justifyContent: 'center', alignItems: 'center', gap: '4px' }, children: [_jsx(Info, { size: 10 }), "Restarting the app is unavaliable. Set \"dat\" in the Authtics Host config to set restriction and then restart."] })) : (_jsx(_Fragment, {})), _jsx(PanelButton, { onClick: () => { shutDown(); }, fullWidth: true, style: { color: '#fff', justifyContent: 'center', height: '32px', background: '#ba4e4e' }, disabled: datRestrictionMissing, children: _jsx("span", { style: { display: 'flex', alignItems: 'center', gap: '6px', textAlign: 'center' }, children: "Shut Down App" }) }), datRestrictionMissing ? (_jsxs("span", { style: { display: 'flex', opacity: 0.5, fontSize: 10, justifyContent: 'center', alignItems: 'center', gap: '4px' }, children: [_jsx(Info, { size: 10 }), "Shutting down the app is unavaliable. Set \"dat\" in the Authtics Host config to set restriction and then restart."] })) : (_jsx(_Fragment, {})), _jsx(PanelButton, { onClick: () => { pause_Resume_Users(); }, fullWidth: true, style: { color: '#fff', justifyContent: 'center', height: '32px', background: '#ba4e4e' }, disabled: datRestrictionMissing, children: _jsx("span", { style: { display: 'flex', alignItems: 'center', gap: '6px', textAlign: 'center' }, children: usersPausedRealInfo ? "Resume Users" : "Pause Users" }) }), datRestrictionMissing ? (_jsxs("span", { style: { display: 'flex', opacity: 0.5, fontSize: 10, justifyContent: 'center', alignItems: 'center', gap: '4px' }, children: [_jsx(Info, { size: 10 }), "Pausing users from accessing the app is unavaliable. Set \"dat\" in the Authtics Host config to set restriction and then restart."] })) : (_jsx(_Fragment, {}))] })] })) }) }), datMatch ? (_jsxs("div", { onClick: () => setOpen(o => !o), style: {
                            display: 'flex',
                            flexDirection: 'row',
                            alignItems: 'center',
                            padding: '8px 12px',
                            borderRadius: hasIssue ? '100px' : '10px',
                            background: hasIssue ? '#ba4e4e' : '#111',
                            border: hasIssue ? '1px solid #991b1b' : '1px solid #333',
                            boxShadow: hasIssue
                                ? '0 4px 20px rgba(220,38,38,0.2)'
                                : '0 4px 20px rgba(0,0,0,0.4)',
                            color: '#fff',
                            cursor: 'pointer',
                            transition: 'all 0.4s ease',
                            width: 'fit-content',
                            boxSizing: 'border-box',
                            lineHeight: '1',
                        }, children: [_jsx(ShieldCheck, { size: 16, color: "#f472b6", style: { flexShrink: 0 } }), _jsx("div", { style: {
                                    maxWidth: hasIssue ? '160px' : '0px',
                                    opacity: hasIssue ? 1 : 0,
                                    overflow: 'hidden',
                                    whiteSpace: 'nowrap',
                                    marginLeft: hasIssue ? '8px' : '0px',
                                    transition: 'max-width 0.4s ease, opacity 0.4s ease, margin-left 0.4s ease',
                                }, children: _jsxs("span", { style: {
                                        display: 'inline-flex',
                                        alignItems: 'center',
                                        gap: '4px',
                                        background: 'rgba(255,255,255,0.15)',
                                        padding: '3px 8px',
                                        borderRadius: '100px',
                                        fontSize: '12px',
                                    }, children: [_jsx(CircleX, { size: 11 }), errorCount > 0 && _jsxs("span", { children: [errorCount, " error", errorCount > 1 ? 's' : '', "."] })] }) }), _jsx("div", { style: { marginLeft: hasIssue ? '6px' : '4px', opacity: 0.4, display: 'flex' }, children: open ? _jsx(ChevronDown, { size: 12 }) : _jsx(ChevronUp, { size: 12 }) })] })) : (_jsx(_Fragment, {}))] })] }));
}
