// @bananacool467/authtics-host/dist/renderEngine.js
import path from "node:path";
import fs from "node:fs";
import { fileURLToPath } from "node:url";
import * as esbuild from "esbuild";
import * as babel from "@babel/core";
import * as babelPresetEnv from "@babel/preset-env";
import React from "react";
import { renderToString } from "react-dom/server";
import { tempFileForTsx, writeTempFile, readTempFile, ensureTempDirs, } from "./rendererTempFiles.js";
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
async function buildWithBabel(entryPath, virtualContent, platform) {
    const presets = [
        [babelPresetEnv, {
                targets: platform === 'node' ? { node: 'current' } : { browsers: '>0.5%, not dead' },
                modules: platform === 'node' ? 'commonjs' : false,
            }],
    ];
    const result = await babel.transformAsync(virtualContent, {
        filename: entryPath,
        presets,
        sourceMaps: false,
    });
    if (!result?.code)
        throw new Error('[authtics] Babel produced no output');
    return result.code;
}
async function buildWithEsbuild(virtualContent, resolveDir, platform) {
    const isNode = platform === 'node';
    const result = await esbuild.build({
        stdin: {
            contents: virtualContent,
            resolveDir,
            sourcefile: `virtual-entry-${platform}.jsx`,
        },
        bundle: true,
        write: false,
        format: isNode ? 'cjs' : 'iife',
        platform,
        jsx: 'automatic',
        sourcemap: isNode ? false : false,
        target: isNode ? ['node18'] : ['es2020'],
        minify: true,
        treeShaking: true,
        external: isNode
            ? ['react', 'react-dom', 'react-dom/server', 'react/jsx-runtime', 'react/jsx-dev-runtime', 'node:http',
                'node:path',
                'node:fs',
                'node:url',
                'tls',
                path.join(__dirname, 'cli.js'),
                path.join(__dirname, 'config.js'),
                path.join(__dirname, 'renderEngine.js'),
                path.join(__dirname, 'rendererTempFiles.js')]
            : [
                path.join(__dirname, 'cli.js'),
                path.join(__dirname, 'cli.js'),
                path.join(__dirname, 'config.js'),
                path.join(__dirname, 'renderEngine.js'),
                path.join(__dirname, 'rendererTempFiles.js')
            ],
    });
    return result.outputFiles[0].text;
}
export function findFileUpwards(startDir, relativePath) {
    let dir = path.resolve(startDir);
    const root = path.parse(dir).root;
    while (true) {
        const candidate = path.join(dir, relativePath);
        if (fs.existsSync(candidate))
            return candidate;
        if (dir === root)
            break;
        dir = path.dirname(dir);
    }
    return null;
}
function resolveDevPanelPath() {
    const nextToCli = path.join(__dirname, "devPanel.js");
    if (fs.existsSync(nextToCli))
        return nextToCli;
    return null;
}
const devPanelPath = resolveDevPanelPath();
function resolveTemplate(appDir, autoPageDir, autoPaging) {
    let nextToCli;
    if (autoPaging) {
        nextToCli = path.join(process.cwd(), autoPageDir || "", "_template.html");
    }
    else {
        nextToCli = path.join(process.cwd(), appDir || "", "../", "_template.html");
    }
    return nextToCli;
}
function buildHtml(opts) {
    const { pageHtml, clientCode, devPanelContainer, templatePath, hotReload, fileName } = opts;
    const hmrScript = hotReload ? HMR_SCRIPT : '';
    if (templatePath && fs.existsSync(templatePath)) {
        let template = fs.readFileSync(templatePath, "utf-8");
        const contentTag = /<authtics_host_content\s*\/>/i;
        if (contentTag.test(template)) {
            template = template.replace(contentTag, `<div id="root">${pageHtml}</div>\n    ${devPanelContainer}`);
            template = template.includes("</body>")
                ? template.replace("</body>", `  ${fileName ? `<script src="${fileName}"></script>` : `<script>${clientCode}</script>`}\n${hmrScript}</body>`)
                : template + `\n${fileName ? `<script src="${fileName}"></script>` : `<script>${clientCode}</script>`}\n${hmrScript}`;
            return template;
        }
        console.warn('[authtics] _template.html has no <authtics_host_content /> — using default layout.');
    }
    return `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
  </head>
  <body>
    <div id="root">${pageHtml}</div>
    ${devPanelContainer}
    ${fileName ? `<script src="${fileName}"></script>` : `<script>${clientCode}</script>`}
    ${hmrScript}
  </body>
</html>`;
}
function isCacheValid(tsxPath, tempPath, serverCjsPath, templatePath, devPanelPath) {
    if (!fs.existsSync(tempPath) || !fs.existsSync(serverCjsPath))
        return false;
    try {
        const cacheMtime = fs.statSync(tempPath).mtimeMs;
        if (fs.statSync(tsxPath).mtimeMs > cacheMtime)
            return false;
        if (templatePath && fs.existsSync(templatePath) && fs.statSync(templatePath).mtimeMs > cacheMtime)
            return false;
        if (devPanelPath && fs.existsSync(devPanelPath) && fs.statSync(devPanelPath).mtimeMs > cacheMtime)
            return false;
        return true;
    }
    catch {
        return false;
    }
}
const HMR_SCRIPT = `<script>(function(){const es=new EventSource('/hmr');es.onmessage=function(e){if (e.data === 'reload') window.location.reload()};es.onerror=function(){setTimeout(function(){window.location.reload()},1000)};})();</script>`;
export async function renderPage(tsxPath, reqUrl, showPanel = true, compiler = 'esbuild', hotReload = false, autoPaging = true, autoPageDir, appDir, fileHandler, rmvFileHandler, reqDAT, configDAT) {
    ensureTempDirs();
    if (!fs.existsSync(tsxPath)) {
        throw new Error(`renderPage: source file not found: ${tsxPath}`);
    }
    let usesDevPanel = showPanel && devPanelPath !== null;
    const templatePath = resolveTemplate(appDir, autoPageDir, autoPaging);
    const tempPath = tempFileForTsx(tsxPath);
    const serverCjsPath = tempPath.replace(/\.json$/, ".server.cjs");
    let clientCode;
    let serverCode;
    if (isCacheValid(tsxPath, tempPath, serverCjsPath, templatePath, devPanelPath) && autoPaging) {
        try {
            const parsed = JSON.parse(readTempFile(tempPath));
            clientCode = parsed.clientCode;
            serverCode = fs.readFileSync(serverCjsPath, "utf-8");
        }
        catch {
        }
    }
    if (!clientCode || !serverCode) {
        const virtualServer = `
      ${usesDevPanel ? `import DevPanel from ${JSON.stringify(devPanelPath)};` : ""}
      import Page from ${JSON.stringify(tsxPath)};
      export default Page;
      ${usesDevPanel ? `export { DevPanel };` : ""}
    `;
        const virtualClient = `
      import React from "react";
      import { hydrateRoot } from "react-dom/client";
      ${usesDevPanel ? `import DevPanel from ${JSON.stringify(devPanelPath)};` : ""}
      import Page from ${JSON.stringify(tsxPath)};

      ${usesDevPanel ? `
      const panelEl = document.getElementById("authtics-dev-panel-root");
      if (panelEl) hydrateRoot(panelEl, React.createElement(DevPanel));
      ` : ""}
      const rootEl = document.getElementById("root");
      if (rootEl) hydrateRoot(rootEl, React.createElement(Page));
    `;
        if (compiler === 'babel') {
            serverCode = await buildWithBabel(tsxPath, virtualServer, 'node');
            clientCode = await buildWithBabel(tsxPath, virtualClient, 'browser');
        }
        else {
            serverCode = await buildWithEsbuild(virtualServer, process.cwd(), 'node');
            clientCode = await buildWithEsbuild(virtualClient, process.cwd(), 'browser');
        }
        fs.writeFileSync(serverCjsPath, serverCode, "utf-8");
        writeTempFile(tempPath, JSON.stringify({ clientCode }));
    }
    fileHandler ? fileHandler(clientCode, `/${tempPath.substring(tempPath.lastIndexOf('/') + 1).replace(".json", ".js").replace(".tsx_", ".").toLowerCase()}`) : null;
    const serverModule = await import(`${serverCjsPath}?t=${Date.now()}`);
    const exports = serverModule.default ?? serverModule;
    const PageComponent = exports.default ?? exports.Page ?? exports.page ?? null;
    const DevPanelComponent = usesDevPanel ? (exports.DevPanel ?? null) : null;
    if (!PageComponent) {
        throw new Error(`renderPage: no default export found in ${tsxPath}`);
    }
    const pageHtml = renderToString(React.createElement(PageComponent));
    const devPanelHtml = DevPanelComponent
        ? renderToString(React.createElement(DevPanelComponent))
        : "";
    const devPanelContainer = usesDevPanel
        ? `<div id="authtics-dev-panel-root">${devPanelHtml}</div>`
        : "";
    return buildHtml({ pageHtml, clientCode, devPanelContainer, templatePath, hotReload, fileName: `/${tempPath.substring(tempPath.lastIndexOf('/') + 1).replace(".json", ".js").replace(".tsx_", ".").toLowerCase()}` });
}
